File: 0139/dfs.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// XXX: TLE
class Solution {
 public:
  bool wordBreak(string s, vector<string>& wordDict) {
    return dfs(0, s, wordDict);
  }

  bool dfs(size_t start_idx, string& s, vector<string>& wordDict) {
    if (start_idx == s.size()) return true;
    for (auto& word : wordDict) {
      if (s.substr(start_idx, word.size()) == word) {
        bool ret = dfs(start_idx + word.size(), s, wordDict);
        if (ret) return true;
      }
    }
    return false;
  }
};

int main() {
  Solution sol;
  string s;
  vector<string> word_dict;

  s = "leetcode";
  word_dict = {"leet", "code"};
  assert(sol.wordBreak(s, word_dict) == true);

  s = "applepenapple";
  word_dict = {"apple", "pen"};
  assert(sol.wordBreak(s, word_dict) == true);

  s = "catsandog";
  word_dict = {"cats", "dog", "sand", "and", "cat"};
  assert(sol.wordBreak(s, word_dict) == false);

  s = "cars";
  word_dict = {"cars", "ca", "rs"};
  assert(sol.wordBreak(s, word_dict) == true);

  s = "abcd";
  word_dict = {"a", "abc", "b", "cd"};
  assert(sol.wordBreak(s, word_dict) == true);
}

File: 1405/dfs.py
from typing import *

# XXX: TLE
class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        return self.dfs(a, b, c, "", {})

    def dfs(self, a: int, b: int, c: int, s: str, memo: Dict[Tuple, str]) -> str:
        ret = s
        if a > 0 and (len(s) <= 1 or s[-2:] != "aa"):
            candidate = self.dfs(a - 1, b, c, s + "a", memo)
            if len(candidate) > len(ret):
                ret = candidate
        if b > 0 and (len(s) <= 1 or s[-2:] != "bb"):
            candidate = self.dfs(a, b - 1, c, s + "b", memo)
            if len(candidate) > len(ret):
                ret = candidate
        if c > 0 and (len(s) <= 1 or s[-2:] != "cc"):
            candidate = self.dfs(a, b, c - 1, s + "c", memo)
            if len(candidate) > len(ret):
                ret = candidate
        return ret


if __name__ == '__main__':
    sol = Solution()

    a, b, c = 1, 1, 7
    expected = ("ccaccbcc", "ccbccacc")
    assert sol.longestDiverseString(a, b, c) in expected

    a, b, c = 7, 1, 0
    expected = "aabaa"
    assert sol.longestDiverseString(a, b, c) == expected
